### PR TITLE
Refactor pipeline nodes

### PR DESCRIPTION
**Pull request by mariakleiner**
use map instead of array

### CODE REVIEW

**Comment by mariakleiner** [comment]
nodeTypes `key` -> `id` and `name` -> `displayName` rename coming up next

File: pkg/Library/NodeGraph/CandidateFinder.js [hunk 1]
```diff
@@ -50,9 +51,9 @@ findConnectionCandidates(storeName, {$type}, node, {nodes}, nodeTypes, globalSto
   const candidates = [];
   candidates.push(this.findGlobalCandidate(storeName, globalStores));
   
-  nodes.forEach(n => {
-    if (n.key !== node.key) {
-      candidates.push(this.findCandidateInNode(n.key, $type, nodeTypes[n.type]));
+  keys(nodes).forEach(key => {
```
Comment (sjmiles) [inline pkg/Library/NodeGraph/CandidateFinder.js:54]: Think this is easier if you iterate the `values`.

File: pkg/Library/NodeGraph/ConnectionUpdator.js [hunk 2]
```diff
@@ -28,19 +27,23 @@ update({pipeline, nodeTypes, candidates}, state) {
   }
 },
 
+inputsChanged({pipeline, candidates}, state) {
+  return pipeline &&
```
Comment (sjmiles) [inline pkg/Library/NodeGraph/ConnectionUpdator.js:31]: I would not have noticed this probably except by doing code review, but the contract of `pipelineChanged` is different between this module and `CandidateFinder.js`. It's a micro-nit. :)

File: pkg/Library/NodeGraph/Editor.js [hunk 3]
```diff
@@ -13,15 +13,11 @@ update({pipeline, selectedNodeKey}, state) {
     state.selectedPipelineName = pipeline?.$meta.name;
     // new pipeline, choose a selectedNode if there isn't one
     if (!selectedNodeKey) {
-      return {selectedNodeKey: pipeline?.nodes?.[0]?.key};
+      return {selectedNodeKey: keys(pipeline?.nodes)?.[0]?.key};
```
Comment (sjmiles) [inline pkg/Library/NodeGraph/Editor.js:16]: I think this is not quite right? The zeroth key is `keys(pipeline?.nodes)[0]` or `keys(pipeline?.nodes.pop())`.

File: pkg/Library/NodeGraph/NodeInspector.js [hunk 4]
```diff
@@ -28,15 +28,15 @@ async update(inputs, state, {service, output, invalidate}) {
       // calling async method without awaiting it, on purpose,
       // which creates Special Circumstances, see below
       this.monitorStores(state, nodeTypes, {service, invalidate});
-    }
+    }
```
Comment (sjmiles) [inline pkg/Library/NodeGraph/NodeInspector.js:31]: WS? My editor will delete it is all, so it will be in a commit one way or another :)

**Review by sjmiles — CHANGES_REQUESTED** [review]
Great!

There is one actual bug I found, and the rest are nits or notes.

General note: consider looking for places where we are sending 'pipeline' but only using 'pipeline.nodes'. It may be possible to write simpler 'nodes functions' instead of 'pipeline functions'.

File: pkg/Library/NodeGraph/CandidateFinder.js (hunk 1 above)
Comment (mariakleiner) [reply pkg/Library/NodeGraph/CandidateFinder.js]: done

File: pkg/Library/NodeGraph/ConnectionUpdator.js (hunk 2 above)
Comment (mariakleiner) [reply pkg/Library/NodeGraph/ConnectionUpdator.js:31]: good catch (thanks for the NodesConnector refactoring and breaking it into 3 separate particles, in 2 of these, the pipelineChanged implementation could be significantly simplified - work only needed, if switching to another pipeline, or node(s) added/removed)

File: pkg/Library/NodeGraph/Editor.js (hunk 3 above)
Comment (mariakleiner) [reply pkg/Library/NodeGraph/Editor.js]: fixed, thank you!

File: pkg/Library/NodeGraph/NodeInspector.js (hunk 4 above)
Comment (mariakleiner) [reply pkg/Library/NodeGraph/NodeInspector.js]: fixed 
PS: what is "WS"?

**Comment by mariakleiner** [comment]
addressed comments PTAL

File: pkg/Library/NodeGraph/NodeInspector.js (hunk 4 above)
Comment (sjmiles) [reply pkg/Library/NodeGraph/NodeInspector.js]: whitespace

**Review by sjmiles — APPROVED** [review]

**Pull request merged**